### PR TITLE
BUG: ArrowExtensionArray.mode(dropna=False) not respecting NAs

### LINKED ISSUE #50982: BUG: .mode with pyarrow dtypes ignores dropna
```python
ser = pd.Series([1, 1, None, None], dtype="int64[pyarrow]")

ser.mode(dropna=True)
ser.mode(dropna=False)
```

With either dropna=True or dropna=False, we get back `Series([1], dtype="int64[pyarrow]")`.  I'd expect that with dropna=True, but expeced to get `Series([1, None], dtype="int64[pyarrow]")` with dropna=False.

Encountered while trying to fix the mode xfail for string dtype by implementing it in terms of value_counts, which does behave the way I expected.

### PR DESCRIPTION
**Pull request by mroeschke**
- [x] closes #50982 (Replace xxxx with the GitHub issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/core/arrays/arrow/array.py [hunk 1]
```diff
@@ -1269,7 +1269,6 @@ def _mode(self: ArrowExtensionArrayT, dropna: bool = True) -> ArrowExtensionArra
         ----------
         dropna : bool, default True
             Don't consider counts of NA values.
-            Not implemented by pyarrow.
```
Comment (jbrockmendel) [inline pandas/core/arrays/arrow/array.py:1272]: worth asking pyarrow to implement this upstream?

File: pandas/core/arrays/arrow/array.py [hunk 2]
```diff
@@ -1292,15 +1291,21 @@ def _mode(self: ArrowExtensionArrayT, dropna: bool = True) -> ArrowExtensionArra
             data = self._data
 
         modes = pc.mode(data, pc.count_distinct(data).as_py())
-        values = modes.field(0)
         counts = modes.field(1)
         # counts sorted descending i.e counts[0] = max
+        if not dropna and self._data.null_count > counts[0].as_py():
```
Comment (jbrockmendel) [inline pandas/core/arrays/arrow/array.py:1296]: i know you've wanted to keep as much of this "in pyarrow" as possible, but i find this hard to follow.  would it be that bad to just implement mode in terms of value_counts?

```
if not len(self):[...]
vcs = self.value_counts(dropna=dropna)
res_ser = vcs[vcs == vcs.max()].sort_index()
return res_ser.index._values
```

**Comment by jbrockmendel** [comment]
looks like this breaks the existing mode tests

**Comment by jbrockmendel** [comment]
Couple comments, neither deal-breakers.  Otherwise LGTM

File: pandas/core/arrays/arrow/array.py (hunk 1 above)
Comment (mroeschke) [reply pandas/core/arrays/arrow/array.py:1326]: Sure opened https://github.com/apache/arrow/issues/34173

**Comment by jbrockmendel** [comment]
@jorisvandenbossche can you take a look?  this calls a bunch of pyarrow stuff directly

File: pandas/core/arrays/arrow/array.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/arrays/arrow/array.py:1347]: makes sense, thanks for taking a look

File: pandas/core/arrays/arrow/array.py [hunk 3]
```diff
@@ -1343,15 +1342,21 @@ def _mode(self: ArrowExtensionArrayT, dropna: bool = True) -> ArrowExtensionArra
...
+            return type(self)(pa.array([None], type=pa_type))
```
Comment (jorisvandenbossche) [inline pandas/core/arrays/arrow/array.py:1348]: I just wanted to comment this on the arrow issue that you could do something like the above as a workaround, but so you are already doing that;) 
Personally, I think it's a fine workaround on the pandas side (certainly on the short term). It should also be basically as performant compared to when pyarrow would do this in `mode` itself, since `null_count` is cheap (and cached).

File: pandas/core/arrays/arrow/array.py (hunk 2 above)
Comment (jorisvandenbossche) [reply pandas/core/arrays/arrow/array.py:1347]: Seeing this comment here, I find it surprising that there is such a big difference between both. Especially because you are doing a count_distinct to include all unique values in the result of `mode`, so essentially making it equivalent to a value_counts. So if there is a big difference, that seems to indicate some performance issue in the implementation in arrow. 

Now, trying with the following, I see different results:

```
In [8]: arr = pa.array(np.random.randint(0, 1000, 1_000_000))

In [12]: %timeit pc.mode(arr, pc.count_distinct(arr).as_py())
12.7 ms ± 496 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [13]: %timeit pc.value_counts(arr)
8.96 ms ± 170 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```
(it might depend a lot on the exact characteristics of the data, though, number of uniques vs total number of values, etc)

Comment (mroeschke) [reply pandas/core/arrays/arrow/array.py:1347]: This was a quick benchmark replacing the implementation to use `ArrowExtensionArray.value_counts`, not just comparing `pc.mode` vs `pc.value_counts` per se

```
% ipython
Python 3.8.15 | packaged by conda-forge | (default, Nov 22 2022, 08:53:40)
Type 'copyright', 'credits' or 'license' for more information
IPython 8.8.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import pyarrow as pa

In [2]: data = list(range(100_000)) + [None] * 100_000

In [3]: arr = pd.arrays.ArrowExtensionArray(pa.array(data))

In [4]: %timeit arr._mode()
15.9 ms ± 57.7 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

% git diff
+        vcs = self.value_counts(dropna=dropna)
+        res_ser = vcs[vcs == vcs.max()].sort_index()
+        return res_ser.index._values
+        # pa_type = self._data.type
+        # if pa.types.is_temporal(pa_type):
+        #     nbits = pa_type.bit_width
+        #     if nbits == 32:
+        #         data = self._data.cast(pa.int32())
+        #     elif nbits == 64:
+        #         data = self._data.cast(pa.int64())
+        #     else:
+        #         raise NotImplementedError(pa_type)
+        # else:
+        #     data = self._data
+        #
+        # modes = pc.mode(data, pc.count_distinct(data).as_py())
+        # counts = modes.field(1)
+        # # counts sorted descending i.e counts[0] = max
+        # if not dropna and self._data.null_count > counts[0].as_py():
+        #     return type(self)(pa.array([None], type=pa_type))
+        # mask = pc.equal(counts, counts[0])
+        # most_common = modes.field(0).filter(mask)
+        #
+        # if pa.types.is_temporal(pa_type):
+        #     most_common = most_common.cast(pa_type)
+        #
+        # if not dropna and self._data.null_count == counts[0].as_py():
+        #     most_common = pa.concat_arrays(
+        #         [most_common, pa.array([None], type=pa_type)]
+        #     )
+        #
+        # return type(self)(most_common)

% ipython
Python 3.8.15 | packaged by conda-forge | (default, Nov 22 2022, 08:53:40)
Type 'copyright', 'credits' or 'license' for more information
IPython 8.8.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import pyarrow as pa

In [2]: data = list(range(100_000)) + [None] * 100_000

In [3]: arr = pd.arrays.ArrowExtensionArray(pa.array(data))

In [4]: %timeit arr._mode()
54.7 ms ± 672 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

Comment (jorisvandenbossche) [reply pandas/core/arrays/arrow/array.py:1347]: Then I suppose the overhead is coming from the extra work the `ArrowExtensionArray.value_counts` is doing? 

Doing it with pyarrow compute directly here in `_mode` might be faster and simpler (compared to current `_mode`)? Something like: 

```
res = pc.value_counts(self.data)
most_common = res.field("values").filter(pc.equal(res.field("counts"), pc.max(res.field("counts"))))
```

(this is still faster than calling `pc.mode` on your example data)

File: pandas/tests/extension/test_arrow.py [hunk 4]
```diff
@@ -1372,19 +1372,12 @@ def test_quantile(data, interpolation, quantile, request):
 
 @pytest.mark.parametrize(
     "take_idx, exp_idx",
-    [[[0, 0, 2, 2, 4, 4], [4, 0]], [[0, 0, 0, 2, 4, 4], [0]]],
+    [[[0, 0, 2, 2, 4, 4], [0, 4]], [[0, 0, 0, 2, 4, 4], [0]]],
     ids=["multi_mode", "single_mode"],
 )
 def test_mode_dropna_true(data_for_grouping, take_idx, exp_idx, request):
     pa_dtype = data_for_grouping.dtype.pyarrow_dtype
-    if pa.types.is_string(pa_dtype) or pa.types.is_binary(pa_dtype):
```
Comment (jbrockmendel) [inline pandas/tests/extension/test_arrow.py:1381]: nice!

File: pandas/core/arrays/arrow/array.py (hunk 2 above)
Comment (mroeschke) [reply pandas/core/arrays/arrow/array.py]: Nice! Looks like `pc.value_counts` also has some benefits

1. Works for string and binary types
2. The result maintains the order of the original values.

So it's good to switch to using value_counts here

File: pandas/core/arrays/arrow/array.py [hunk 5]
```diff
@@ -1345,12 +1344,13 @@ def _mode(self: ArrowExtensionArrayT, dropna: bool = True) -> ArrowExtensionArra
...
             data = self._data
 
-        modes = pc.mode(data, pc.count_distinct(data).as_py())
-        values = modes.field(0)
-        counts = modes.field(1)
-        # counts sorted descending i.e counts[0] = max
-        mask = pc.equal(counts, counts[0])
-        most_common = values.filter(mask)
+        if dropna:
+            data = data.drop_null()
```
Comment (jorisvandenbossche) [inline pandas/core/arrays/arrow/array.py:1348]: I don't know if you checked, but it might be more efficient to do this after the value_counts, so on `res` (assuming that `res` is a much shorter array, and so cheaper to filter)

Comment (mroeschke) [reply pandas/core/arrays/arrow/array.py:1348]:
```
In [1]: import pyarrow as pa

In [2]: data = list(range(100_000)) + [None] * 100_000

In [3]:  arr = pd.arrays.ArrowExtensionArray(pa.array(data))

In [4]: %timeit arr._mode()
7.01 ms ± 148 µs per loop (mean ± std. dev. of 7 runs, 100 loops each) <-- drop_null before
6.93 ms ± 112 µs per loop (mean ± std. dev. of 7 runs, 100 loops each) <-- drop_null after
```

So might as well do this after as you suggested

Comment (mroeschke) [reply pandas/core/arrays/arrow/array.py:1363]: Looks like filtering after gives incorrect result for multi-mode tests. If filtering were to occur after, I would have to drop the NAs in `values` and then filter the `counts` where the NA were in `values`. To keep things simpler for now I'll give leave this to filter before the `value_counts`

Comment (jorisvandenbossche) [reply pandas/core/arrays/arrow/array.py:1363]: > If filtering were to occur after, I would have to drop the NAs in `values` and then filter the `counts` where the NA were in `values`

That would be something like:

```
if dropna:
    res = res.filter(res.field("values").is_valid())
```

to drop values based on one field of the struct, before calculating `most_common`.

So that line is a bit more complicated as calling `drop_null`, but only slightly. Now, it also doesn't seem to matter that much ;)

Comment (mroeschke) [reply pandas/core/arrays/arrow/array.py:1363]: Ah thanks. This passes the tests but appears slower than dropping the NAs beforehand for this example, so I think we should just drop the NAs beforehand for now.

```
In [1]: import pyarrow as pa

In [2]: data = list(range(100_000)) + [None] * 100_000

In [3]: arr = pd.arrays.ArrowExtensionArray(pa.array(data))

In [4]: %timeit arr._mode()
6.72 ms ± 45 µs per loop (mean ± std. dev. of 7 runs, 100 loops each) <- drop_nulls before
7.24 ms ± 87 µs per loop (mean ± std. dev. of 7 runs, 100 loops each) <- fllter, is_valild
```

**Comment by mroeschke** [comment]
Greenish

**Review by jbrockmendel — APPROVED** [review]
LGTM thanks for taking this on
